### PR TITLE
Internet connectivity in MicroVMs

### PR DESCRIPTION
**Pull request by ustiugov**
* Added a test for internet connectivity
* Set up nameservers when creating a VM
* Added functionality to configure iptables at the tap creation

Signed-off-by: Dmitrii Ustiugov <dmitrii.ustiugov@epfl.ch>

### CODE REVIEW

**Comment by ustiugov** [comment]
need to add the iptables commands from the wiki to 
https://github.com/ease-lab/vhive/blob/f43f531d79f12253f7cd26b51537a1daa0937f69/taps/tapManager.go#L104

need to drop the first commit that reduces the number of bridges

**Review by shyamjesal — APPROVED** [review]

File: ctriface/Makefile [hunk 1]
```diff
@@ -28,6 +28,11 @@ WITHLAZY:=-lazy
 GOBENCH:=-v -timeout 1500s
 CTRDLOGDIR:=/tmp/ctrd-logs
 
+test-ping:
```
[the remark this answers is not in the document]
Comment (shyamjesal) [reply ctriface/Makefile:31]: removed.

File: taps/tapManager.go [hunk 2]
```diff
@@ -100,6 +104,89 @@ func createBridge(bridgeName, gatewayAddr string) {
...
+		log.Warnf("Failed to fetch net interfaces %v\n%s\n", err, out)
+	}
+	scanner := bufio.NewScanner(bytes.NewReader(out))
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.Contains(line, "default") {
+			ethernetID = line[strings.LastIndex(line, " ")+1:]
+		}
+	}
+	cmd := exec.Command(
```
[the remark this answers is not in the document]
Comment (shyamjesal) [reply taps/tapManager.go:124]: done

File: taps/tapManager.go [hunk 3]
```diff
@@ -100,6 +104,89 @@ func createBridge(bridgeName, gatewayAddr string) {
...
+		"-m",
+		"conntrack",
+		"--ctstate",
+		"RELATED,ESTABLISHED",
+		"-j",
+		"ACCEPT",
+	)
+	stdoutStderr, err = cmd.CombinedOutput()
+	if err != nil {
+		log.Warnf("Failed to configure conntrack %v\n%s\n", err, stdoutStderr)
```
[the remark this answers is not in the document]
Comment (shyamjesal) [reply taps/tapManager.go:186]: added.

File: ctriface/iface.go [hunk 4]
```diff
@@ -306,6 +306,7 @@ func (o *Orchestrator) getVMConfig(vm *misc.VM) *proto.CreateVMRequest {
 				IPConfig: &proto.IPConfiguration{
 					PrimaryAddr: vm.Ni.PrimaryAddress + vm.Ni.Subnet,
 					GatewayAddr: vm.Ni.GatewayAddress,
+					Nameservers: []string{"10.96.0.10", "8.8.8.8"},
```
[the remark this answers is not in the document]
Comment (shyamjesal) [reply ctriface/iface.go:309]: K9s DNS is dynamic now.

File: ctriface/iface.go [hunk 5]
```diff
@@ -291,6 +293,21 @@ func (o *Orchestrator) getImage(ctx context.Context, imageName string) (*contain
 func (o *Orchestrator) getVMConfig(vm *misc.VM) *proto.CreateVMRequest {
 	kernelArgs := "ro noapic reboot=k panic=1 pci=off nomodules systemd.log_color=false systemd.unit=firecracker.target init=/sbin/overlay-init tsc=reliable quiet 8250.nr_uarts=0 ipv6.disable=1"
 
+	//using googleDNS as a backup
```
Comment (ustiugov) [inline ctriface/iface.go:296]: please make it a helper function

File: taps/tapManager.go [hunk 6]
```diff
@@ -100,6 +104,89 @@ func createBridge(bridgeName, gatewayAddr string) {
 	}
 }
 
+//ConfigIPtables Configures IP tables for internet access inside VM
+func ConfigIPtables(tapName string) {
+
+	var ethernetID string
```
[the remark this answers is not in the document]
Comment (shyamjesal) [reply taps/tapManager.go]: done

**Review by ustiugov — COMMENTED** [review]
lgtm

**Review by shyamjesal — APPROVED** [review]

**Pull request merged**